src/uploads/imageResizer.ts

```diff
@@ -6,128 +6,212 @@ import sharp from 'sharp';
   const extension = sourceImage.split('.').pop();
   const name = sanitize(sourceImage.substring(0, sourceImage.lastIndexOf('.')) || sourceImage);
-
-  return {
-    name,
-    extension,
-    width: size.width,
-    height: size.height,
-  };
-}
-
+  return { name, ext: extension! };
+};
+
+/**
+ * Create a new image name based on the output image name, the dimensions and
+ * the extension.
+ *
+ * Ignore the fact that duplicate names could happen if the there is one
+ * size with `width AND height` and one with only `height OR width`. Because
+ * space is expensive, we will reuse the same image for both sizes.
+ *
+ * @param sanitizedImage - the sanitized image name
+ * @param bufferInfo - the buffer info
+ * @param extension - the extension to use
+ * @param takenOutputNames - the taken output names
+ * @returns the new image name that is not taken
+ */
+const createImageName = (
+  outputImageName: string,
+  { width, height }: sharp.OutputInfo,
+  extension: string,
+) => `${outputImageName}-${width}x${height}.${extension}`;
+
+/**
+ * Create the result object for the image resize operation based on the
+ * provided parameters. If the name is not provided, an empty result object
+ * is returned.
+ *
+ * @param name - the name of the image
+ * @param filename - the filename of the image
+ * @param width - the width of the image
+ * @param height - the height of the image
+ * @param filesize - the filesize of the image
+ * @param mimeType - the mime type of the image
+ * @param sizesToSave - the sizes to save
+ * @returns the result object
+ */
+const createResult = (
+  name: string,
+  filename: FileSize['filename'] = null,
+  width: FileSize['width'] = null,
+  height: FileSize['height'] = null,
+  filesize: FileSize['filesize'] = null,
+  mimeType: FileSize['mimeType'] = null,
+  sizesToSave: FileToSave[] = [],
+): ResizeResult => ({
+  sizesToSave,
+  sizeData: {
+    [name]: {
+      width,
+      height,
+      filename,
+      filesize,
+      mimeType,
+    },
+  },
+});
+
+/**
+ * Check if the image needs to be resized according to the requested dimensions
+ * and the original image size. If the resize options are provided, the image
+ * will be resized regardless of the requested dimensions, given that the
+ * width or height to be resized is provided.
+ *
+ * @param requestedDimensions - the requested dimensions
+ * @param original - the original image size
+ * @param resizeOptions - the resize options
+ * @returns true if the image needs to be resized, false otherwise
+ */
+const needsResize = (
+  { width, height, withoutEnlargement, withoutReduction }: ImageSize,
+  original: ProbedImageSize,
+): boolean => {
+  // allow enlargement or prevent reduction (our default is to prevent
+  // enlargement and allow reduction)
+  if (withoutEnlargement === false || withoutReduction) {
+    return true;
+  }
+
+  const isWidthOrHeightDefined = !!height || !!width;
+  const hasInsufficientWidth = !!width && width <= original.width;
```

Comment:
Shouldn't this be the other way around? So

`  const hasInsufficientWidth = !!width && original.width <= width;`

Reply:
I kept the version that was in place and only added variable names.

```ts
 return (typeof desiredSize.width === 'number' && desiredSize.width <= dimensions.width)
    || (typeof desiredSize.height === 'number' && desiredSize.height <= dimensions.height)
    || (!desiredSize.height && !desiredSize.width);
```